### ISSUE TITLE
Websocket get the wrong path in a request with express and shiny server proxy

### ISSUE THREAD

**Issue by appernetic** [nodejitsu/node-http-proxy#1005]
I am trying to proxy shiny server with node.js. I can do that with no problem but if I am trying to access it within an angular app  the web socket request is wrong. 

I have added base url in my angular view:
```
<base href="http://localhost:8070/api/">
<base href="ws://localhost:8070/api/">
```
The ws base url makes no difference in solving the issue.

When I access the URL in the browser: http://localhost:8070/api 

It works and the url the shiny app use is correct.

It looks like this:
http://localhost:8070/api/__sockjs__/n=U4KBxHv7R0QfzjfwJR/info

When I access the URL in my Angular controller like this: 
```
$http({
  method: 'GET',
  url: 'http://localhost:8070/api/',
}).then(function successCallback(response) {
$scope.theView = response.data;
  }, function errorCallback(response) {
    $scope.theView = response.data;
  });
```
I output the scope in the view:
```
 <div ng-bind-html="theView | to_trusted" class="container-fluid text-left"></div>
```
All URL’s is correct except the websocket request (/api/ is missing):

http://localhost:8070/__sockjs__/n=quKu5RbSqQ5oKNRVGa/info

In console you can also see   Referer:http://localhost:8070/

Because of this I get the error message: ”Disconnected from the server”

In the shiny server I have both tried to put my R code in: 

/srv/shiny-server/api

and /srv/shiny-server/

No difference in behavior. It works if I access the url directly but not with Angular.

In node.js I use node-http-proxy and express for routing:
```
var express  = require('express');
var app      = express();
var httpProxy = require('http-proxy');
var apiProxy = httpProxy.createProxyServer({ws: true, prependPath:true, target: {
    host: 'localhost',
    path: '/api/',
    port: 3838
  }});

// angular app
app.use('/', express.static(__dirname + '/'));

// proxy
app.use('/api/',
  function(req, res) {

  res.setHeader("Content-Type", "charset=utf-8");
  
    apiProxy.web(req, res); 

    // websocket upgrade
    apiProxy.on('upgrade', function (req, socket, head) {
      proxy.ws(req, socket, head);
    });

    // error
    apiProxy.on('error', function (err, req, res) {
      res.writeHead(500, {
        'Content-Type': 'text/plain'
      });
      res.end('Proxy error. Plase try again in a moment. Error='+err);
      throw err;
    });

    // close
    apiProxy.on('close', function (req, socket, head) {
      console.log('Client disconnected');
    });
});



app.listen(8070, function (err) {
  console.log('listening:  ' + '8070');
});
```
My shiny code is the example hello shiny:
```
library(shiny)

# Define UI for application that draws a histogram
shinyUI(fluidPage(

  # Application title
  titlePanel("Hello Shiny!"),

  # Sidebar with a slider input for the number of bins
  sidebarLayout(
    sidebarPanel(
      sliderInput("bins",
                  "Number of bins:",
                  min = 1,
                  max = 50,
                  value = 30)
    ),

    # Show a plot of the generated distribution
    mainPanel(
      plotOutput("distPlot")
    )
  )
))
```

What could cause this issue?
Could it be the express configuration?

**Comment by appernetic** [comment]
I have thought about the issue and I think that if I could change the referer for the websocket request It should work. If it is possible to do it in the angular http function or in the node server somewhere...

**Comment by appernetic** [comment]
It doesn't work to add a header to $http service in angular. I have also tried to add a header in node.js:

res.setHeader("Referer", "http://localhost:8070/api/");

No luck so far.

**Comment by appernetic** [comment]
Nor is the browser happy if you change the referer header in Angular. 

To trick the browser and WebSocket (actually socks) that the origin is from the same path as in the shiny app I first enabled html5Mode in angular and then did a route rewriting function in the node server. 

Then all paths was wrong except the WebSocket path wich was working. So I had to modify the shiny app also. Now it's working. 

All issues was client side. No changes in the proxy.

Now on to next problem authenticating socks with auth0.

Closing!
